### ISSUE TITLE
Multi module project: Building only a sub module does not resolve versions for deps

### ISSUE THREAD

**Issue by jschneider** [jgitver/jgitver-maven-plugin#21]
Multi module setup. 
Defining dependencyManagement of module versions in root pom.xml

When building only one sub module, the version for dependencies to other sub modules is not updated.

E.g.:

root/
    module1
    module2

"module2" depends on "module1". The version is defined as "${project.version}" in the root/pom.xml.
When building in "root", everything works as expected.

But when I change the directory to "module2" and call "mvn install", Maven tries (and fails) to download 

Could not resolve dependencies for project module2:jar:3.14.0-SNAPSHOT: The following artifacts could not be resolved: module2:jar:0

**Comment by xeagle2** [comment]
@jschneider, @McFoggy your issue has the same root cause as https://github.com/jgitver/jgitver-maven-plugin/issues/10#issuecomment-231529228 and there would be much more issues with the current approach (I have already faced more than three critical issues) how this module works. 

I'm not the author of the plugin, however I really like the idea of the plugin itself to dynamically calculate the version. However due to the specific behavior of maven's internal container and model building process, most of the interpolation and initialization stuff have been done before the plugin starts making any calculation and modifications, therefore most of the functionality with complex configuration wouldn't work as expected and build would be broken.

Another issue is how the plugin overwrites the model persisting the changes to temporary .pom files and resets the pom location for the project after that. The plugin assumes and reads the model from original .pom file

`Model model = loadInitialModel(project.getFile());`

modifies the version

`model.setVersion(newProjectVersions.get(initalProjectGAV));`

persists the changes to temporary .pom

```
File newPom = createPomDumpFile();
writeModelPom(model, newPom);
```

and resets the pom location for project

`setProjectPomFile(project, newPom);`

I personally don't think this method is conceptually correct, because at the moment when plugin loads the model from local .pom file, the model in memory might contain much more modifications and these modifications are completely lost. 

Additionally it looks like the resetting the pom file for project operation doesn't give any effect at least for latest versions of maven (found that it helps to have adjusted .pom installed to local repository)

So, my thoughts are that the conceptually the plugin should be adjusted, otherwise there would be much more issues and broken builds discovered. At the moment I'm testing one more new option of the integration and if the proof of the concept would bring expected results, hopefully it would be possible to get it properly working at least for latest versions of maven.

**Comment by McFoggy** [comment]
@xeagle2 I fully agree that for maven the current approach does not suit maven pom loading internals which are done and thus interpreted way too much early.

For maven, we need to step back and think of other alternatives I agree. If you have ideas feel free to share via email or here behind an issue, thanks for your interest.

**Comment by xeagle2** [comment]
@McFoggy, @jschneider could you please try to build the version based on my modifications https://github.com/xeagle2/jgitver-maven-plugin/tree/xeagle2 (branch called 'xeagle2'), try to connect the plugin as described at https://github.com/xeagle2/jgitver-maven-plugin/tree/xeagle2#as-native-maven-extension-experimental.

If it works as expected for you, we would need to merge the changes to main upstream making the Pull Request.

Thank you,

**Comment by McFoggy** [comment]
@xeagle2 the usage of the ModelProcessor looks better/safer than the current implementation. On the other hand I am not a big fan of the serialization/deserialization of the WorkingConfiguration.

**Comment by xeagle2** [comment]
@McFoggy, I'm not a fun of the object serialization either, however there is no any other quick way to transfer the data  within the session avoiding the primitive types as strings, etc. using the different class loaders. Xpp3Dom unfortunately is either not an option due the class loading schema. So, any other way you can find to be properly functional, always welcome to integrate :) however the currently experimental solution works much better and more stable in more complex build configurations.

**Comment by McFoggy** [comment]
@xeagle2 I'd like to introduce your changes regarding model handling but without (at least in a first step)  the pom changes & attachement via dynamic plugin and so forth. Can I cherry pick some of your changes  in a branch? or can you simplify your work in some branch and do a PR if you prefer?

and do not bother too much with backward compatibility, we'll break it ; we are not yet in any final version so it is still time.

**Comment by xeagle2** [comment]
@McFoggy, I would be able to compile the changes in a separate branch and make a PR afterwards. I'm a bit confused with your words regarding the pom changes, etc. I didn't modify pom, I introduced the changes in Java so far. It would be nice if we can clarify this.

**Comment by McFoggy** [comment]
this issue should now be corrected with the 0.3.0-alpha2 release.
Please review the README to use the new maven core extension mechnism.
